<pre>
<center><b><h1>Bank Marketing Campaign</b></center>
<center><b><h1>Week - 5</b></center>
<pre>    

<pre>
<center><b><h1>Model Evaluation</b></center>
<pre>  

### Step 1. Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import joblib

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    classification_report,
    confusion_matrix
)

### Step 2. Assign dataset to a variable called df

In [2]:
df = pd.read_csv("bank-full.csv",sep=";")
df

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45206,51,technician,married,tertiary,no,825,no,no,cellular,17,nov,977,3,-1,0,unknown,yes
45207,71,retired,divorced,primary,no,1729,no,no,cellular,17,nov,456,2,-1,0,unknown,yes
45208,72,retired,married,secondary,no,5715,no,no,cellular,17,nov,1127,5,184,3,success,yes
45209,57,blue-collar,married,secondary,no,668,no,no,telephone,17,nov,508,4,-1,0,unknown,no


### Step 3: Separate X and y

In [3]:
X = df.drop('y', axis=1)
y = df['y']

### Step 4: Convert target yes/no into 1/0

In [4]:
y = y.map({'no': 0, 'yes': 1})

### Step 5: Identify numerical and categorical columns

In [5]:
numeric_columns = [
    'age',
    'balance',
    'day',
    'duration',
    'campaign',
    'pdays',
    'previous'
]

print("Numerical columns:")
print(numeric_columns)

Numerical columns:
['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


In [6]:
categorical_columns = X.select_dtypes(
    include=['object']
).columns.tolist()

print("\nCategorical columns:")
print(categorical_columns)


Categorical columns:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome']


### Step 6: Convert categorical columns into numbers

In [7]:
X = pd.get_dummies(X, columns=categorical_columns, drop_first=True)

In [8]:
X = X.astype(int)

### Step 7: Split the data into training and testing

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

### Step 8: Scale numerical features

In [10]:
scaler = StandardScaler()

X_train[numeric_columns] = scaler.fit_transform(
    X_train[numeric_columns]
)

X_test[numeric_columns] = scaler.transform(
    X_test[numeric_columns]
)

In [11]:
X_train

,age,balance,day,duration,campaign,pdays,previous,job_blue-collar,job_entrepreneur,job_housemaid,...,month_jul,month_jun,month_mar,month_may,month_nov,month_oct,month_sep,poutcome_other,poutcome_success,poutcome_unknown
24001,-0.460434,-0.164410,1.582124,-0.457311,-0.246104,-0.410910,-0.241509,0,0,0,...,0,0,0,0,0,0,0,0,0,1
43409,-1.589641,0.899627,-1.298384,2.502493,0.398202,1.446096,2.664584,0,0,0,...,0,0,0,0,0,0,0,0,0,0
20669,0.292371,-0.365486,-0.458236,5.697691,0.398202,-0.410910,-0.241509,0,0,0,...,0,0,0,0,0,0,0,0,0,1
18810,0.668773,-0.445003,1.822166,-0.862499,2.653271,-0.410910,-0.241509,0,0,0,...,1,0,0,0,0,0,0,0,0,1
23130,-0.272233,-0.361249,1.222060,-0.777602,2.331118,-0.410910,-0.241509,0,0,0,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
17958,0.856974,-0.146160,1.702145,-0.773743,-0.246104,-0.410910,-0.241509,1,0,0,...,1,0,0,0,0,0,0,0,0,1
15941,-0.460434,-0.437834,0.741975,-0.700423,0.720354,-0.410910,-0.241509,0,0,0,...,1,0,0,0,0,0,0,0,0,1
16952,0.386471,-0.419258,1.102039,-0.619386,-0.568256,-0.410910,-0.241509,1,0,0,...,1,0,0,0,0,0,0,0,0,1
34781,-1.307339,0.388954,-1.178363,-0.121583,-0.568256,-0.410910,-0.241509,0,0,0,...,0,0,0,1,0,0,0,0,0,1


### Step 9 : Load Existing Model(.pkl file)

In [12]:
model = joblib.load("model.pkl")

print("Existing model loaded successfully!")
print(model)

Existing model loaded successfully!
LogisticRegression(max_iter=1000)


### Step 10 : Make Predictions Using Existing model

In [13]:
y_pred = model.predict(X_test)

print("\nFirst 20 predictions:")
print(y_pred[:20])


First 20 predictions:
[0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]


### Step 11 : Model Evaluation

In [14]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Model Evaluation")
print("----------------")
print("Accuracy :", accuracy)
print("Precision:", precision)
print("Recall   :", recall)
print("F1 Score :", f1)

Model Evaluation
----------------
Accuracy : 0.9015813336282207
Precision: 0.6473684210526316
Recall   : 0.3487712665406427
F1 Score : 0.4533169533169533


### Step 12 : Classification Report

In [15]:
print("Classification Report")
print("---------------------")

print(classification_report(
    y_test,
    y_pred,
    target_names=["No", "Yes"]
))

Classification Report
---------------------
              precision    recall  f1-score   support

          No       0.92      0.97      0.95      7985
         Yes       0.65      0.35      0.45      1058

    accuracy                           0.90      9043
   macro avg       0.78      0.66      0.70      9043
weighted avg       0.89      0.90      0.89      9043



### Step 13 : Confusion Matrix

In [16]:
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[7784  201]
 [ 689  369]]


### Step 14 : Check for Underfitting / Overfitting

In [17]:
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)

print("Overfitting / Underfitting Check")
print("--------------------------------")
print("Training Accuracy:", train_score)
print("Testing Accuracy :", test_score)

difference = train_score - test_score

print("Difference       :", difference)

Overfitting / Underfitting Check
--------------------------------
Training Accuracy: 0.9020128290201282
Testing Accuracy : 0.9015813336282207
Difference       : 0.0004314953919075659


In [18]:
if train_score > 0.90 and difference > 0.10:
    print("Result: Model may be overfitting.")

elif train_score < 0.70 and test_score < 0.70:
    print("Result: Model may be underfitting.")

else:
    print("Result: Model appears to be fitting reasonably well.")

Result: Model appears to be fitting reasonably well.


### Step 15 — 5-Fold Cross Validation

In [19]:
cv_scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=5,
    scoring='accuracy'
)

print("5-Fold Cross Validation Scores:")
print(cv_scores)

print("Mean CV Accuracy:", cv_scores.mean())
print("Standard Deviation:", cv_scores.std())

5-Fold Cross Validation Scores:
[0.89894941 0.90102295 0.89977882 0.90501866 0.90349786]
Mean CV Accuracy: 0.9016535393011196
Standard Deviation: 0.002278162632487183


### Step 16 — Try Advanced Classification Models

In [20]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_test)

In [21]:
print("Random Forest")
print("Accuracy :", accuracy_score(y_test, rf_pred))
print("Precision:", precision_score(y_test, rf_pred))#Precision = TP / (TP + FP)
print("Recall   :", recall_score(y_test, rf_pred))#Recall = TP / (TP + FN)
print("F1 Score :", f1_score(y_test, rf_pred))

Random Forest
Accuracy : 0.9044564856795311
Precision: 0.655448717948718
Recall   : 0.38657844990548207
F1 Score : 0.4863258026159334


### Step 17 — AdaBoost

In [22]:
from sklearn.ensemble import AdaBoostClassifier

ada_model = AdaBoostClassifier(
    n_estimators=100,
    random_state=42
)

ada_model.fit(X_train, y_train)

ada_pred = ada_model.predict(X_test)

In [23]:
print("AdaBoost")
print("Accuracy :", accuracy_score(y_test, ada_pred))
print("Precision:", precision_score(y_test, ada_pred))
print("Recall   :", recall_score(y_test, ada_pred))
print("F1 Score :", f1_score(y_test, ada_pred))

AdaBoost
Accuracy : 0.8976003538648678
Precision: 0.6157894736842106
Recall   : 0.33175803402646503
F1 Score : 0.4312039312039312


### Step 18 — Gradient Boosting

In [24]:
from sklearn.ensemble import GradientBoostingClassifier

gb_model = GradientBoostingClassifier(
    n_estimators=100,
    random_state=42
)

gb_model.fit(X_train, y_train)

gb_pred = gb_model.predict(X_test)

In [25]:
print("Gradient Boosting")
print("Accuracy :", accuracy_score(y_test, gb_pred))
print("Precision:", precision_score(y_test, gb_pred))
print("Recall   :", recall_score(y_test, gb_pred))
print("F1 Score :", f1_score(y_test, gb_pred))

Gradient Boosting
Accuracy : 0.9043459029083268
Precision: 0.6459909228441755
Recall   : 0.40359168241965976
F1 Score : 0.4968004653868528


### Step 19 — Make ONE Model Comparison Table

Model	            Accuracy	Precision	 Recall	  F1-score
Logistic Regression	0.9016   	0.6474	     0.3488	   0.4533
Random Forest	    0.9044      0.6554       0.3865    0.4863
AdaBoost	        0.8976      0.6157       0.3317    0.4312
Gradient Boosting	0.9043      0.6459       0.4035    0.4968

### Step 21 — Hyperparameter Tuning

In [26]:
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2]
}

In [27]:
grid_search = GridSearchCV(
    RandomForestClassifier(random_state=42),
    param_grid,
    cv=5,
    scoring='f1',
    n_jobs=-1
)

grid_search.fit(X_train, y_train)

print("Best Parameters:")
print(grid_search.best_params_)

print("Best CV Score:")
print(grid_search.best_score_)

Best Parameters:
{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Best CV Score:
0.49145200736407996


In [29]:
from sklearn.neighbors import KNeighborsClassifier

import warnings
warnings.filterwarnings("ignore")
# Create KNN model
knn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model
knn_model.fit(X_train, y_train)

# Make predictions
knn_pred = knn_model.predict(X_test)

# Evaluate KNN
print("KNN")
print("Accuracy :", accuracy_score(y_test, knn_pred))
print("Precision:", precision_score(y_test, knn_pred))
print("Recall   :", recall_score(y_test, knn_pred))
print("F1 Score :", f1_score(y_test, knn_pred))

KNN
Accuracy : 0.8960521950680084
Precision: 0.5930599369085173
Recall   : 0.3553875236294896
F1 Score : 0.4444444444444444


In [30]:
sample = X_test.iloc[[0]]

prediction = knn_model.predict(sample)

if prediction[0] == 1:
    print("Prediction: Yes")
else:
    print("Prediction: No")

Prediction: No


### Compare all 5 models

In [31]:
print("\nModel Comparison")
print("----------------")

print("Logistic Regression :", accuracy_score(y_test, y_pred))
print("Random Forest       :", accuracy_score(y_test, rf_pred))
print("AdaBoost            :", accuracy_score(y_test, ada_pred))
print("Gradient Boosting   :", accuracy_score(y_test, gb_pred))
print("KNN                 :", accuracy_score(y_test, knn_pred))


Model Comparison
----------------
Logistic Regression : 0.9015813336282207
Random Forest       : 0.9044564856795311
AdaBoost            : 0.8976003538648678
Gradient Boosting   : 0.9043459029083268
KNN                 : 0.8960521950680084
